File: leet100/02_double_pointers/leet_03_15.cpp
#include <vector>
#include <algorithm>

#include "../../cpp/utils/str_tools.h"

// 超时
class Solution1 {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums)
    {
        std::sort(nums.begin(), nums.end());

        std::set<std::vector<int>> ss;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > 0) {
                break;
            }
            for (int j = i+1; j < nums.size(); j++) {
                int diff = 0 - nums[i] - nums[j];
                auto it = std::find(nums.begin() + j + 1, nums.end(), diff);
                if (it != nums.end()) {
                    ss.insert({nums[i], nums[j], *it});
                }
            }
        }   
        
        std::vector<std::vector<int>> results(ss.begin(), ss.end());
        return results;
    }
};

class Solution {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums)
    {   
        std::vector<std::vector<int>> results;

        std::sort(nums.begin(), nums.end());
        

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > 0) {
                break;
            }
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }

            int right = nums.size() - 1;
            for (int j = i+1; j < nums.size(); j++) {
                if (j > i+1 && nums[j] == nums[j-1]) {
                    continue;
                }

                while (j < right && nums[i] + nums[j] + nums[right] > 0) {
                    right--;
                }

                if (j == right) {
                    break;
                }

                if (nums[i] + nums[j] + nums[right] == 0) {
                    results.push_back({nums[i], nums[j], nums[right]});
                }
            }
        }

        return results;
    }
};

void test1()
{
    Solution sn;
    std::vector<int> nums {-1, 0, 1, 2, -1, -4};
    std::vector<std::vector<int>> result = sn.threeSum(nums);
    std::cout << "results:" << std::endl;
    for (auto& vec : result) {
        std::cout << StringTools::ToString(vec) << std::endl;
    }
    std::cout << "expected: [[-1,-1,2],[-1,0,1]]" << std::endl;
}


void test2()
{
    Solution sn;
    std::vector<int> nums {0, 1, 1};
    std::vector<std::vector<int>> result = sn.threeSum(nums);
    std::cout << "results:" << std::endl;
    for (auto& vec : result) {
        std::cout << StringTools::ToString(vec) << std::endl;
    }
    std::cout << "expected: []" << std::endl;
}

void test3()
{
    Solution sn;
    std::vector<int> nums {0, 0, 0};
    std::vector<std::vector<int>> result = sn.threeSum(nums);
    std::cout << "results:" << std::endl;
    for (auto& vec : result) {
        std::cout << StringTools::ToString(vec) << std::endl;
    }
    std::cout << "expected: [[0,0,0]]" << std::endl;
}

int main()
{
    test1();
    test2();
    test3();
    return 0;
}

File: leet100/02_double_pointers/leet_03_15.py
from typing import List

# 会超时
class Solution1:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)

        result: List[List[int]] = []
        for i in range(len(nums)):
            if nums[i] > 0:  # 提前退出
                break
            for j in range(i+1, len(nums)):
                diff = 0 - (nums[i] + nums[j])
                if diff in nums[j+1:]:
                    sorted_tmp = [nums[i], nums[j], diff]
                    # sorted_tmp = sorted(sorted_tmp)  # 由于在一开始就sorted过了，所以这里没必要
                    if sorted_tmp not in result:
                        result.append([nums[i], nums[j], diff])
        return result

# 会超时
class Solution2:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        results = set()
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                to_find = 0 - (nums[i] + nums[j])
                for k in range(j+1, len(nums)):
                    if nums[k] == to_find:
                        item = [nums[i], nums[j], nums[k]]
                        results.add(tuple(sorted(item)))  # 借助tuple，set才能去重
                        break

        return [list(t) for t in results]   # 把tuple转成list

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []

        nums.sort()
        for first in range(len(nums)):
            if first > 0 and nums[first] == nums[first-1]:
                continue

            third = len(nums) - 1
            for second in range(first+1, len(nums)):
                if second > first + 1 and nums[second] == nums[second-1]:
                    continue

                while second < third and nums[first] + nums[second] + nums[third] > 0:
                    third -= 1

                if second == third:
                    break

                if nums[first] + nums[second] + nums[third] == 0:
                    result.append([nums[first], nums[second], nums[third]])
        return result

def test1():
    s = Solution()
    nums = [-1, 0, 1, 2, -1, -4]
    result = s.threeSum(nums)
    print(f"result:{result}, expected: [[-1, -1, 2], [-1, 0, 1]]")

def test2():
    s = Solution()
    nums = [0, 1, 1]
    result = s.threeSum(nums)
    print(f"result:{result}, expected: []")

def test3():
    s = Solution()
    nums = [0,0,0]
    result = s.threeSum(nums)
    print(f"result:{result}, expected: [[0,0,0]]")

if __name__ == "__main__":
    test1()
    test2()
    test3()